src/measure/dump_dipole.cu

```diff
@@ -0,0 +1,237 @@
+/*
+    Copyright 2017 Zheyong Fan, Ville Vierimaa, Mikko Ervasti, and Ari Harju
+    This file is part of GPUMD.
+    GPUMD is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+    GPUMD is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details. You should have received a copy of the GNU General
+   Public License along with GPUMD.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*-----------------------------------------------------------------------------------------------100
+Dump energy/force/virial with all loaded potentials at a given interval.
+--------------------------------------------------------------------------------------------------*/
+
+#include "dump_dipole.cuh"
+#include "force/nep3.cuh"
+#include "model/box.cuh"
+#include "model/read_xyz.cuh"
+#include "parse_utilities.cuh"
+#include "utilities/common.cuh"
+#include "utilities/error.cuh"
+#include "utilities/gpu_vector.cuh"
+#include "utilities/read_file.cuh"
+#include <iostream>
+#include <vector>
+
+__device__ void warpReduceDipole(volatile float* sdata, int tid)
```

Comment:
This is deprecated (and in principle wrong) way to do reduction. You can follow the style of the similar kernels I wrote (there is one in force.cu). 

Saem comment for polarizability.

Reply:
Ah okay, I see! I followed the one in force.cu, it was much shorter and neater :D